src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/HtmlCSharp/OnTypeRenameHandlerTest.cs

```diff
@@ -0,0 +1,278 @@
+            {
+                LanguageKind = RazorLanguageKind.Html,
+            };
+            var projectionProvider = GetProjectionProvider(projectionResult);
+
+            var documentMappingProvider = GetDocumentMappingProvider(expectedResponse.Ranges, 0 /* Different from document version (1) */, RazorLanguageKind.Html);
+
+            var onTypeRenameHandler = new OnTypeRenameHandler(documentManager, requestInvoker, projectionProvider, documentMappingProvider);
+            var onTypeRenameRequest = new DocumentOnTypeRenameParams()
+            {
+                TextDocument = new TextDocumentIdentifier() { Uri = Uri },
+                Position = new Position(10, 5)
+            };
+
+            // Act
+            var result = await onTypeRenameHandler.HandleRequestAsync(onTypeRenameRequest, new ClientCapabilities(), CancellationToken.None).ConfigureAwait(false);
+
+            // Assert
+            Assert.True(called);
+            Assert.Null(result);
+        }
+
+        [Fact]
+        public async Task HandleRequestAsync_RemapFailure_DiscardsLocation()
+        {
+            // Arrange
+            var called = false;
+            var expectedResponse = GetMatchingHTMLBracketRange(5);
+            var documentManager = new TestDocumentManager();
+            documentManager.AddDocument(Uri, Mock.Of<LSPDocumentSnapshot>(d => d.Version == 0));
+
+            var htmlResponse = GetMatchingHTMLBracketRange(10);
+            var requestInvoker = GetRequestInvoker<DocumentOnTypeRenameParams, DocumentOnTypeRenameResponseItem>(
+                htmlResponse,
+                (method, serverContentType, highlightParams, ct) =>
+                {
+                    Assert.Equal(MSLSPMethods.OnTypeRenameName, method);
+                    Assert.Equal(RazorLSPConstants.HtmlLSPContentTypeName, serverContentType);
+                    called = true;
+                });
+
+            var projectionResult = new ProjectionResult()
+            {
+                LanguageKind = RazorLanguageKind.Html,
+            };
+            var projectionProvider = GetProjectionProvider(projectionResult);
+
+            var documentMappingProvider = Mock.Of<LSPDocumentMappingProvider>();
+
+            var onTypeRenameHandler = new OnTypeRenameHandler(documentManager, requestInvoker, projectionProvider, documentMappingProvider);
+            var onTypeRenameRequest = new DocumentOnTypeRenameParams()
+            {
+                TextDocument = new TextDocumentIdentifier() { Uri = Uri },
+                Position = new Position(10, 5)
+            };
+
+            // Act
+            var result = await onTypeRenameHandler.HandleRequestAsync(onTypeRenameRequest, new ClientCapabilities(), CancellationToken.None).ConfigureAwait(false);
+
+            // Assert
+            Assert.True(called);
+            Assert.Null(result);
+        }
+
+        private LSPProjectionProvider GetProjectionProvider(ProjectionResult expectedResult)
```

Comment:
I should steal from you a bit here, I tend to go a bit too hard on DRY and end up with TestHelper methods that have switch statements and whatnot. I think you've got a good balance of DRY and clean code with these.